### ISSUE TITLE
Result of aggregation depends on vars used in return statement

### ISSUE THREAD

**Issue by lazdmx** [neo4j/neo4j#5385]
Is it correct that query

```
match (n:Unknown)
return collect(n)
```

returns a single row with an empty array, whereas

```
with 42 as x
match (n:Unknown)
return collect(n), x
```

returns nothing?

**Comment by cleishm** [comment]
Hi @lazutkin,

That definitely seems strange! I'll pass that along to our team.

Thanks,
Chris

**Comment by jexp** [comment]
it's intentional. 

_pure_ aggregations (with nothing else returned) on empty results (no rows) return the default (0 for numeric aggregations and an empty list for collect)

**Comment by lazdmx** [comment]
Hi @jexp

I think that behaviour of aggregation should be consistent and shouldn't depend on context (presence of other variables in with/return statement). Also as consequence it doesn't allow to collect data in one request like:

```
match ...
with collect(..) as a

match ...
with a, collect(...) as b

match ...
with a, b, collect(...) as c
....
```

and we need to split a query in set of subqueries and run them serially (since usually results of `n-1` step are arguments to  step `n`) which lead us to bad performance

**Comment by systay** [comment]
@lazutkin Cypher handles grouping by the "context", as you call it. In SQL, you are explicit about how the grouping for an aggregating query is handled using the `GROUP BY` clause. In contrast, Cypher handles this implicitly by looking at the expressions in your query. When you mix columns with aggregating functions and columns without, the ones without aggregations are used as the grouping keys.

In this particular case, since the non-aggregating column is a literal, you could argue that Cypher could allow it to be used without interpreting is as a grouping key. We have decided that it's easier to have a simple rule (does it contain aggregation) when deciding what is a grouping key and what is a aggregating column.

TL;DR; This is by design.
